### PR TITLE
Add CBMC proof for DNS

### PR DESCRIPTION
**Pull request by tony-josi-aws**
Description
-----------
This PR adds CBMC proofs for the following functions:

- ulNBNSHandlePacket
- prepareReplyDNSMessage
- prvFindEndPointOnNetMask
- DNS_TreatNBNS
- prvCreateDNSMessage

Test Steps
-----------
```
cd test/cbmc/proof/
python3 prepare.py
cd <test_case>
make
```

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "list.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_DNS.h"
+#include "FreeRTOS_IP_Private.h"
+
+#include "cbmc.h"
+
+size_t prvCreateDNSMessage( uint8_t * pucUDPPayloadBuffer,
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c:13]: Why is this re-declaration needed?

File: test/cbmc/proofs/DNS/CreateDNSMessage/Makefile.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "ENTRY": "CreateDNSMessage",
+  "MAX_HOSTNAME_LEN": 5,
+  "HOSTNAME_UNWIND": "__eval {MAX_HOSTNAME_LEN} + 1",
+  "USE_CACHE": 1,
+  "CBMCFLAGS":
+  [
+    "--unwind 1",
+    "--unwindset strlen.0:{HOSTNAME_UNWIND}",
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/CreateDNSMessage/Makefile.json:9]: Good use of keywords.

File: test/cbmc/proofs/DNS/DNSTreatNBNS/DNS_TreatNBNS_harness.c [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+NetworkBufferDescriptor_t * pxResizeNetworkBufferWithDescriptor( NetworkBufferDescriptor_t * pxNetworkBuffer,
+                                                                 size_t xNewSizeBytes )
+{
+    NetworkBufferDescriptor_t * pxRBuf;
+
+    __CPROVER_assert( pxNetworkBuffer != NULL, "pxNetworkBuffer: pvBuffer != NULL" );
+
+    uint8_t * pucNewBuffer = malloc( xNewSizeBytes );
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/DNSTreatNBNS/DNS_TreatNBNS_harness.c:39]: It should be safeMalloc

File: test/cbmc/proofs/DNS/NBNSHandlePacket/NBNSHandlePacket_harness.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    __CPROVER_assume( ( xDataSize < ( ipconfigNETWORK_MTU + ipSIZE_OF_ETH_HEADER ) ) );
+    xNetworkBuffer.xDataLength = xDataSize;
+
+
+    xNetworkBuffer.pucEthernetBuffer = safeMalloc( xDataSize );
+
+    /* pucEthernetBuffer being not NULL is pre validated before the call to ulNBNSHandlePacket */
+    __CPROVER_assume( xNetworkBuffer.pucEthernetBuffer != NULL );
+
+    if( nondet_bool() )
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/NBNSHandlePacket/NBNSHandlePacket_harness.c:41]: Why is this needed, we can directly do safeMalloc for xNetworkBuffer.pxEndPoint, it will automatically dd randomisation of NULL or not NULL

File: test/cbmc/proofs/DNS/prepareReplyDNSMessage/prepareReplyDNSMessage_harness.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    return usReturn;
+}
+
+void harness()
+{
+    uint32_t ulIPAddress;
+    uint16_t usLength;
+
+    /* Assume at least 1 valid endpoint */
+    pxNetworkEndPoint_Temp = ( NetworkEndPoint_t * ) malloc( sizeof( NetworkEndPoint_t ) );
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/prepareReplyDNSMessage/prepareReplyDNSMessage_harness.c:97]: It should be safeMalloc

File: test/cbmc/proofs/DNS/prepareReplyDNSMessage/prepareReplyDNSMessage_harness.c [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    BaseType_t xDataSize;
+
+    /* The pucEthernetBuffer is re adjusted to the following size before the
+     * call to prepareReplyDNSMessage by pxResizeNetworkBufferWithDescriptor. If buffer allocation
+     * scheme #1 (BufferAllocation_1.c) is used we assert if the needed size is actually present
+     * in the buffer.  */
+    __CPROVER_assume( ( xDataSize > ( sizeof( UDPPacket_t ) + sizeof( NBNSRequest_t ) + sizeof( NBNSAnswer_t ) - 2 * sizeof( uint16_t ) ) ) );
+
+    xNetworkBuffer.pucEthernetBuffer = malloc( xDataSize );
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/prepareReplyDNSMessage/prepareReplyDNSMessage_harness.c:108]: Same please update to safeMalloc

File: test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c (hunk 1 above)
Comment (tony-josi-aws) [reply test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c:13]: prvCreateDNSMessage is a static function that is made non static just for CBMC proof with headers

Comment (moninom1) [reply test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c:13]: Yeah i think we have decided on exporting static function already as annirudh suggested.

File: test/cbmc/proofs/DNS/NBNSHandlePacket/NBNSHandlePacket_harness.c (hunk 4 above)
Comment (moninom1) [reply test/cbmc/proofs/DNS/NBNSHandlePacket/NBNSHandlePacket_harness.c]: This is not yet resolved

Comment (tony-josi-aws) [reply test/cbmc/proofs/DNS/NBNSHandlePacket/NBNSHandlePacket_harness.c]: I have resolved it as part of [bda9215](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/924/commits/bda921568042ad0da8575b6dff450b2ce5f6f295) commit made yesterday.

File: test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
+/* FreeRTOS includes. */
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/CreateDNSMessage/CreateDNSMessage_harness.c:1]: Copyright missing

File: test/cbmc/proofs/DNS/DNSTreatNBNS/DNS_TreatNBNS_harness.c [hunk 8]
```diff
@@ -0,0 +1,89 @@
+/* FreeRTOS includes. */
```
Comment (moninom1) [inline test/cbmc/proofs/DNS/DNSTreatNBNS/DNS_TreatNBNS_harness.c:1]: Add copyright

**Review by moninom1 — APPROVED** [review]

**Review by shubnil — APPROVED** [review]
